event_listener/src/main/java/org/jetbrains/maven/server/IntellijMavenSpy.java

```diff
@@ -66,86 +68,83 @@ else if (event instanceof DependencyResolutionResult) {
-  }
-
-  private static String getErrorMessage(Exception exception) {
-    String baseMessage = exception.getMessage();
-    Throwable rootCause = ExceptionUtils.getRootCause(exception);
-    String rootMessage = rootCause != null ? rootCause.getMessage() : StringUtils.EMPTY;
-    return  "ERROR=" + rootMessage + "MESSAGE=" + baseMessage;
-  }
-
-  private static void printSessionStartedEventAndReactorData(ExecutionEvent event, String projectId) {
-    MavenSession session = event.getSession();
-    if (session != null) {
-      List<MavenProject> projectsInReactor = session.getProjects();
-      if (projectsInReactor == null) {
-        projectsInReactor = new ArrayList<MavenProject>();
-      }
-      StringBuilder builder = new StringBuilder();
-      for (MavenProject project : projectsInReactor) {
-        builder
-          .append(project.getGroupId()).append(":")
-          .append(project.getArtifactId()).append(":")
-          .append(project.getVersion()).append("&&");
-      }
-      printMavenEventInfo(ExecutionEvent.Type.SessionStarted, "id", projectId, "projects", builder.toString());
+
+    private static void onRepositoryEvent(RepositoryEvent event) {
+        String errMessage = event.getException() == null ? "" : event.getException().getMessage();
+        String path = event.getFile() == null ? "" : event.getFile().getPath();
+        String artifactCoord = event.getArtifact() == null ? "" : event.getArtifact().toString();
+        printMavenEventInfo(event.getType(), "path", path, "artifactCoord", artifactCoord, "error", errMessage);
     }
-    else {
-      printMavenEventInfo(ExecutionEvent.Type.SessionStarted, "id", projectId, "projects", "");
+
+    private static void onExecutionEvent(ExecutionEvent event) {
+        MojoExecution mojoExec = event.getMojoExecution();
+        String projectId = event.getProject() == null ? "unknown" : event.getProject().getId();
+        if (mojoExec != null) {
+            String errMessage = event.getException() == null ? "" : getErrorMessage(event.getException());
+            printMavenEventInfo(event.getType(), "source", mojoExec.getSource(), "goal", mojoExec.getGoal(), "id", projectId, "ERR",
+                    errMessage);
+        } else {
+            if (event.getType() == ExecutionEvent.Type.SessionStarted) {
+                printSessionStartedEventAndReactorData(event, projectId);
+            } else {
+                printMavenEventInfo(event.getType(), "id", projectId);
+            }
+        }
+    }
+
+    private static String getErrorMessage(Exception exception) {
+        String baseMessage = exception.getMessage();
+        Throwable rootCause = ExceptionUtils.getRootCause(exception);
+        String rootMessage = rootCause != null ? rootCause.getMessage() : StringUtils.EMPTY;
+        return "ERROR=" + rootMessage + "MESSAGE=" + baseMessage;
+    }
+
+    private static void printSessionStartedEventAndReactorData(ExecutionEvent event, String projectId) {
```

Comment:
a lot of things is happening there - probably it could be split and simplified using java-specific things without nested ifs and editing one variable

Reply:
Same as above.